src/config.cc

```diff
@@ -341,97 +343,124 @@ void Config::updateFromCommandLine(const boost::program_options::variables_map&
-      sconfig.flasher = "";
-
-      uptane.secondary_configs.push_back(sconfig);
-    }
+    readSecondaryConfigs(sconfigs);
   }
 
   if (cmd.count("legacy-interface") != 0) {
     std::string legacy_interface = cmd["legacy-interface"].as<std::string>();
-    std::stringstream command;
-    std::string output;
-    command << legacy_interface << " api-version --loglevel " << loggerGetSeverity();
-    int rs = Utils::shell(command.str(), &output);
-    if (rs != 0) {
-      LOGGER_LOG(LVL_error, "Legacy external flasher api-version command failed: " << output);
-    } else if (output != "1") {
+    if (checkLegacyVersion(legacy_interface)) {
+      initLegacySecondaries(legacy_interface);
+    }
+  }
+}
+
+void Config::readSecondaryConfigs(const std::vector<std::string>& sconfigs) {
+  std::vector<std::string>::const_iterator it;
+  for (it = sconfigs.begin(); it != sconfigs.end(); ++it) {
+    Json::Value config_json = Utils::parseJSONFile(*it);
+    Uptane::SecondaryConfig sconfig;
+
+    std::string stype = config_json["secondary_type"].asString();
+    if (stype == "virtual") {
+      sconfig.secondary_type = Uptane::kVirtual;
+    } else if (stype == "legacy") {
+      LOGGER_LOG(LVL_error, "Legacy secondaries should be initialized with --legacy-interface.");
+      continue;
+    } else if (stype == "uptane") {
+      sconfig.secondary_type = Uptane::kUptane;
+    } else {
+      LOGGER_LOG(LVL_error, "Unrecognized secondary type: " << stype);
+      continue;
+    }
+    sconfig.ecu_serial = config_json["ecu_serial"].asString();
+    sconfig.ecu_hardware_id = config_json["ecu_hardware_id"].asString();
+    sconfig.partial_verifying = config_json["partial_verifying"].asBool();
+    sconfig.ecu_private_key = config_json["ecu_private_key"].asString();
+    sconfig.ecu_public_key = config_json["ecu_public_key"].asString();
+
+    sconfig.full_client_dir = boost::filesystem::path(config_json["full_client_dir"].asString());
+    sconfig.firmware_path = boost::filesystem::path(config_json["firmware_path"].asString());
+    sconfig.metadata_path = boost::filesystem::path(config_json["metadata_path"].asString());
+    sconfig.target_name_path = boost::filesystem::path(config_json["target_name_path"].asString());
+    sconfig.flasher = "";
+
+    uptane.secondary_configs.push_back(sconfig);
+  }
+}
+
+bool Config::checkLegacyVersion(const std::string& legacy_interface) {
+  bool ret = false;
+  if (!boost::filesystem::exists(legacy_interface)) {
+    LOGGER_LOG(LVL_error, "Legacy external flasher not found: " << legacy_interface);
+    return ret;
+  }
+  std::stringstream command;
+  std::string output;
+  command << legacy_interface << " api-version --loglevel " << loggerGetSeverity();
+  int rs = Utils::shell(command.str(), &output);
+  if (rs != 0) {
+    LOGGER_LOG(LVL_error, "Legacy external flasher api-version command failed: " << output);
+  } else {
+    boost::trim_if(output, boost::is_any_of(", \n\r\t"));
```

Comment:
Are You sure we need to remove comma?

Reply:
Need, no. But I figured it might be useful to catch that. Do you think that's likely to be a bad idea?